odoo/addons/test_new_api/tests/test_new_fields.py

```diff
@@ -2361,3 +2361,258 @@ def test_delete_m2o_reference_records(self):
+        self._unlink_option('bacon', self.MODEL_REQUIRED)
+
+        self.assertEqual(my_rec1.my_selection, 'bar')
+        self.assertEqual(my_rec2.my_selection, 'foo')
+        self.assertEqual(my_rec3.my_selection, 'bar')
+
+    def test_ondelete_multiple_explicit(self):
+        my_rec1 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'pikachu'})
+        my_rec2 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'foo'})
+        my_rec3 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'eevee'})
+
+        self.assertEqual(my_rec1.my_selection, 'pikachu')
+        self.assertEqual(my_rec2.my_selection, 'foo')
+        self.assertEqual(my_rec3.my_selection, 'eevee')
+
+        self._unlink_option('pikachu', self.MODEL_REQUIRED)
+        self._unlink_option('eevee', self.MODEL_REQUIRED)
+
+        self.assertEqual(my_rec1.my_selection, 'foo')
+        self.assertEqual(my_rec2.my_selection, 'foo')
+        self.assertEqual(my_rec3.my_selection, 'bar')
+
+    def test_ondelete_multiple_implicit(self):
+        my_rec1 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'charmander'})
+        my_rec2 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'squirtle'})
+        my_rec3 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'bulbasaur'})
+
+        self.assertEqual(my_rec1.my_selection, 'charmander')
+        self.assertEqual(my_rec2.my_selection, 'squirtle')
+        self.assertEqual(my_rec3.my_selection, 'bulbasaur')
+
+        self._unlink_option('charmander', self.MODEL_REQUIRED)
+        self._unlink_option('squirtle', self.MODEL_REQUIRED)
+        self._unlink_option('bulbasaur', self.MODEL_REQUIRED)
+
+        self.assertEqual(my_rec1.my_selection, 'foo')
+        self.assertEqual(my_rec2.my_selection, 'foo')
+        self.assertEqual(my_rec3.my_selection, 'foo')
+
+    def test_ondelete_callback(self):
+        my_rec1 = self.env[self.MODEL_REQUIRED].create({'my_selection': 'knickers'})
+
+        self.assertEqual(my_rec1.my_selection, 'knickers')
+
+        self._unlink_option('knickers', self.MODEL_REQUIRED)
+
+        self.assertEqual(my_rec1.my_selection, 'foo')
+        self.assertFalse(my_rec1.active)
+
+    def test_non_stored_selection(self):
+        my_rec1 = self.env['test_new_api.model_selection_non_stored'].create({})
+        my_rec1.my_selection = 'foo'
+
+        self.assertEqual(my_rec1.my_selection, 'foo')
+
+        self._unlink_option('foo', 'test_new_api.model_selection_non_stored')
+
+        self.assertFalse(my_rec1.my_selection)
+
+
+@common.tagged('selection_ondelete_advanced')
+class TestSelectionOndeleteAdvanced(common.TransactionCase):
+
+    MODEL_BASE = 'test_new_api.model_selection_base'
+    MODEL_REQUIRED = 'test_new_api.model_selection_base_required'
+
+    def test_ondelete_bad_kw(self):
+        Model = self.registry[self.MODEL_REQUIRED]
+        self.addCleanup(setattr, Model, '__bases__', Model.__bases__)
+        self.addCleanup(self.registry.model_cache.clear)
```

Comment:
You may override `setUp` for this, as it is the same for all tests in this class.
Add a comment: "necessary cleanup for resetting changes on the registry".